jenkinsci/tfs-plugin
# Use collection URI in payloads, if possible

This branch is mostly about eliminating technical debt around the determination of collection URI and improving the robustness of the endpoints.  A lot of the decoding is now done more automatically by Jackson deserialization, leading to stronger typing and less chance of error.  This also made it possible to remove some duplicated code that ended up no longer being used.

Manual testing
--------------

1. Crafted an event payload with an **eventType** of `git.pullrequest.merged` such that it contained a value for the upcoming `resourceContainers/collection/baseUrl`, then used `curl` to POST it to the `/team-events/gitPullRequestMerged` endpoint.
    1. A breakpoint I had in the `GitPushEvent#determineCollectionUri()` method allowed me to verify that the `baseUrl` value would be used if it was found.
    2. The triggered build had all the usual behaviour, such as building the exact commit specified, posting the Jenkins build's status to the pull request and commit in TFS/Team Services, etc.
2. Included the event payload from test 1 above as the content of a `team-event` section in a payload that was then POSTed to the `/team-build/build/` endpoint using `curl`.
    1. A breakpoint I had in the `GitPushEvent#determineCollectionUri()` method allowed me to verify that the `baseUrl` value would be used if it was found.
    2. The triggered build also had all the usual behaviour.
3. I also tested without the upcoming `baseUrl` value and confirmed the collection URI could still be "guessed".

Mission accomplished!
